### DISCUSSION TITLE
How to get the number of tabs?

### DISCUSSION THREAD

**Discussion by antocuni** [kovidgoyal/kitty#6424] [Q&A]
I'm trying to write my own tab style. I would like each tab to have a fixed width, i.e. `total_columns / number_of_tabs`. Basically, something which mimics what you get with gnome-terminal tabs:
![image](https://github.com/kovidgoyal/kitty/assets/344507/057753fc-15e8-4c3c-8b68-77ef9a3fbeff)

I played a bit with `tab_bar.py` but I couldn't find a way to achieve that. I can get the total number of columns in the screen using `screen.columns`, but how do I get the total number of tabs?

**Comment by kovidgoyal** [comment]
See https://github.com/kovidgoyal/kitty/discussions/4447 for lots of examples.

**Reply by antocuni** [reply]
sorry, I should have mentioned that I had already looked at it, but couldn't find the answer to my question.
I have looked at it again but I still don't find any example which somehow finds the total number of tabs.

I even tried to look at kitty's source code. I see that the number of tabs is stored in the `TabBar` object, here:
https://github.com/kovidgoyal/kitty/blob/7d2dc252f14916be1d8d3c2ae7d31c57833b4ce5/kitty/tab_bar.py#L476-L480

but still, it doesn't seem to be accessible from `draw_tab()`.

**Comment by kovidgoyal** [comment]
The tab length algorithm is not customizable, to change that you would need ot modify kitty code, specifically the update() function in tab_bar.py

**Comment by elusian** [comment]
I was having the same problem and I have a question: is get_boss() an heavy call? 
Are there drawbacks from reading `len(get_boss().active_tab_manager.tabs)` inside the `draw_tab` function at each iteration?

Also, is there documentation on what is available inside a custom tab_bar.py?

Note that someone did make an approximately equal length tab implementation (https://github.com/kovidgoyal/kitty/discussions/4447#discussioncomment-6454511), using max_tab_length, it just does not work well if you have a status on the right.

**Comment by kovidgoyal** [comment]
On Tue, Aug 15, 2023 at 05:53:52AM -0700, elusian wrote:
> I was having the same problem and I have a question: is get_boss() an heavy call? 

No, its not.

> Are there drawbacks from reading `len(get_boss().active_tab_manager.tabs)` inside the `draw_tab` function at each iteration?

active_tab_manager can be None. And it is not guaranteed that the
active_tab_manager is the one currently being drawn.

> 
> Also, is there documentation on what is available inside a custom tab_bar.py?

it runs in the same address space as the rest of kitty, so everything is
available.

**Reply by elusian** [reply]
Thank you for the answer.

> active_tab_manager can be None. And it is not guaranteed that the
> active_tab_manager is the one currently being drawn.

Can I ask for examples of when this may happen (if it's not too complicated to show)?

**Comment by kovidgoyal** [comment]
On Wed, Aug 16, 2023 at 02:54:23AM -0700, elusian wrote:
> Thank you for the answer.
> 
> > active_tab_manager can be None. And it is not guaranteed that the
> > active_tab_manager is the one currently being drawn.
> 
> Can I ask for examples of when this may happen (if it's not too complicated to show)?

have multiple kitty OS windows. active_tab_manager will only point to
the one that has focus. But tabs cab be updated in any OS window

**Comment by elusian** [comment]
I'm a bit unfamiliar with kitty's architecture, sorry, are you saying that if I open kitty multiple times from the OS they all share the boss object?

**Comment by kovidgoyal** [comment]
Yes, on macOS. And on Linux if you use --insgle-instance or you open OS
windows using ctrl+shift+n

On Wed, Aug 16, 2023 at 03:33:12AM -0700, elusian wrote:
> I'm a bit unfamiliar with kitty's architecture, sorry, are you saying that if I open kitty multiple times from the OS they all share the boss object?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6424#discussioncomment-6739377
> You are receiving this because you modified the open/close state.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________
